### DISCUSSION TITLE
Out of memory, is it a Kingfisher issue?

### DISCUSSION THREAD

**Discussion by tatsuz0u** [onevcat/Kingfisher#1638] [Q&A] [answered]
<img width="500" alt="image" src="https://user-images.githubusercontent.com/31207151/109374392-fc479a80-78ef-11eb-8114-35d08727eebd.png">

```
ScrollView {
    LazyVStack {
        ForEach(1...351, id: \.self) { num in
            if let resourceURL = Bundle.main.url(forResource: "\(num)", withExtension: "jpg") {
                let provider = LocalFileImageDataProvider(fileURL: resourceURL)

                KFImage(source: .provider(provider))
                    .resizable()
                    .scaledToFit()
            }
        }
    }
}
//        List(1...351, id: \.self) { num in
//            if let resourceURL = Bundle.main.url(forResource: "\(num)", withExtension: "jpg") {
//                let provider = LocalFileImageDataProvider(fileURL: resourceURL)
        
//                KFImage(source: .provider(provider))
//                    .resizable()
//                    .scaledToFit()
//            }
//        }
```

Image size: 67KB ~ 6.7MB
Image amount: 351

It crashed when loading images from local files in LazyVStack or List.

<img width="500" alt="image" src="https://user-images.githubusercontent.com/31207151/109374546-22216f00-78f1-11eb-83ec-1957db72eaa6.png">

I even tried to scroll slowly to let it have enough time to clear memory caches...

**Comment by onevcat** [comment]
The only thing I can imagine, if it is due to Kingfisher, is the strong reference in `ImageBinder` which was a workaround for `@State` lost in the nested view (#1631 and #1618). I am not sure about the internal memory mechanism here, but if SwiftUI keeps those `KFImage` somewhere it might be an issue.

Can you try the `fix/weak-image` branch to see if there is any difference?

Or you can also try to set the maximum amount of memory cache allowed in Kingfisher to check how it is working:

```swift
// Limits Kingfisher memory cache to 100 MB
KingfisherManager.shared.cache.memoryStorage.config.totalCostLimit = 100_000
```

**Reply by tatsuz0u** [reply]
I tried this before but it doesn't work.
``KingfisherManager.shared.cache.memoryStorage.config.totalCostLimit = 1024 * 1024 * 500``

100000 is 100MB?
OK...I also tried 1024 * 1024 so it's fine.

**Reply by onevcat** [reply]
By default, Kingfisher will at most use 25% of your device's maximum memory for cache. The memory cache will be purged automatically when you switch the app to the background or received a memory warning. In most cases, this rule works fine. `totalCostLimit` has a unit of "byte". So `1024 * 1024 * 500` means 500GB and means unlimited in real life.

**Reply by tatsuz0u** [reply]
D**n my maths.
I will feedback soon~

**Reply by tatsuz0u** [reply]
Wait, 1024 bytes is equal to 1 kilobyte, right?
And 1024 kilobytes is equal to 1 megabyte.
Isn't it 500MB?

**Reply by tatsuz0u** [reply]
Feedback:
The ``fix/weak-image`` branch didn't work.
``KingfisherManager.shared.cache.memoryStorage.config.totalCostLimit = 100_000`` either.

**Reply by onevcat** [reply]
Oh! I was wrong with that calculation. You are right it is 500MB 😂 (If you set it to a small number like `100_000`, maybe you will always see `.disk` for [`RetrieveImageResult.cacheType`](https://github.com/onevcat/Kingfisher/blob/master/Sources/General/KingfisherManager.swift#L44) in the completion handler.)

I tried a list with unlimited images but cannot reproduce the crash. The memory usage obviously has an upper limit and everything goes fine (I tried with the view [here in the sample app](https://github.com/onevcat/Kingfisher/blob/master/Demo/Demo/Kingfisher-Demo/SwiftUIViews/SwiftUIList.swift)). So I guess it is quite a special case here.

I noticed that you are loading local images, but I only tried images from the network. Can you try to see if the crash also happens for network images? (You can find 700 images from this URL: https://github.com/onevcat/Kingfisher/blob/master/Demo/Demo/Kingfisher-Demo/SwiftUIViews/SwiftUIList.swift#L51)

**Reply by tatsuz0u** [reply]
Yes, that's because I couldn't find a steady source to test.
But the size of images in your dataset seems to be too small to reproduce...

Try this one :-)
``https://github.com/arakitatsuzou/Wallpaper-Dataset/blob/main/\(index).jpg?raw=true``

**Reply by tatsuz0u** [reply]
Release 6.1.1:
LazyVStack ❌
List ❌

Branch fix:
LazyVStack ❌
List ⭕️

**Reply by onevcat** [reply]
Umm, at least for now I found that the properties of `KFImage` in a `LazyVStack` are not deallocated until the whole stack is removed from the screen. While in a `List`, the view and its properties are released as expected. That is why it is fine in a `List` but not in a `LazyVStack`. The `fix/weak-image` should work for remote images (at least as I tested it does not consume memory unboundedly). Can you also confirm that?

(Personally, as I indicated in a few issues/posts, I really do not suggest the use of the `LazyStack` or `LazyGrid`, which are quite buggy for real use. Or more frankly, the whole SwiftUI is not ready for production at all.)

I didn't yet try the local images and I will see if I have a chance soon. However, Kingfisher should be doing the same thing for either local images or network images for caching. So I can hardly imagine a reason.

**Reply by onevcat** [reply]
And in case if you do not know, for these kinds of huge images, I suggest using `downsampling` before loading and setting it, to get better performance while scrolling as well as caching. Something like:

```swift
KFImage.url(url)
    .downsampling(size: .init(width: 300, height: 300))
    .///
```

**Reply by tatsuz0u** [reply]
I have just finished the test.
Branch ``fix/weak-image`` fixed the memory issue in 6.1.1 for ``List``. Network & Local.

**Reply by tatsuz0u** [reply]
By the way, branch ``fix/weak-image`` still can crash due to the same reason if I write like this.

```
List { 
    ForEach(1...351, id: \.self) { num in
        if let url = URL(string: "https://github.com/arakitatsuzou/Wallpaper-Dataset/blob/main/\(num).jpg?raw=true") {
            KFImage(url)
                .resizable()
                .scaledToFit()
        }
    }
}
```

**Comment by onevcat** [comment, marked as answer]
I tried to copy all your images to an empty project and the memory also bursts when using this code:

```swift
struct ContentView: View {
    var body: some View {
        List {
           ForEach(1...351, id: \.self) { num in
                Image(uiImage: image(n: num))
                 .resizable()
                 .scaledToFit()
           }
        }
}

func image(n: Int) -> UIImage! {
    let p = Bundle.main.path(forResource: "\(n)", ofType: "jpg")!
    return UIImage(contentsOfFile: p)
}
```

![截屏2021-03-01 14 31 22](https://user-images.githubusercontent.com/1019875/109455833-cdb9f300-7a9a-11eb-989c-ce1ba89dfff5.png)

> But as you commented, `List(1..<351)` works fine and the `Image` data get released.

So I do not think it is what an issue can be solved by Kingfisher.

SwiftUI just holds everything for the `Image` and there is no chance to release until you destroy the whole `List` (although the cells of the list are reused, the image itself and its data are not released by SwiftUI).

![截屏2021-03-01 14 26 23](https://user-images.githubusercontent.com/1019875/109455514-2937b100-7a9a-11eb-9a19-bcc9714663a7.png)

Again, as I suggested, you can try:

1. Avoid using the buggy/bad implemented views. (Nesting a `ForEach` in a `List`, using `LazyVStack`, etc)
2. [Load a downsampled version](https://github.com/onevcat/Kingfisher/discussions/1638#discussioncomment-414255) of an image.
3. Wait for the next version of SwiftUI to see if it improves.

**Reply by tatsuz0u** [reply]
Thanks!

I really need LazyVStack because List is too ugly. (Separators and horizontal insets)
Hope SwiftUI 3.0 can fix it.

**Reply by onevcat** [reply]
> I really need LazyVStack because `List` is too ugly. (Separators and horizontal insets)

Yes, I agree. In case if you do not know, you can always customize them by using a `UITableView` appearance trick.

**Reply by tatsuz0u** [reply]
Are you living in GitHub wwww

**Comment by dreampiggy** [comment]
Seems you're using loading and rendering too much of images at the same time, and never release or destory them. This can not been solved by any SDKs :)... I guess you're just building some benchmark demo Apps but not real word Apps...

Consider change your UI structure, with some tips:

1. Limit each image's size by using Kingfisher's transformer or SD's`thumbnailPixelSize`. For example, you can use your visible frame size on the screen to limit the image size. (For example, your iPhone XSMax screen pixels size is 1242x2688, showing 25 GIF images, each one limit the `thumbnailPixelSize` to 248x537 pixels)
2. For animation, don't play animation at the same time, you can play the only visible or small part of animated images, leave others stopping and so that SDK can free up the frame buffer.

Releated issues: https://github.com/SDWebImage/SDWebImageSwiftUI/issues/172

**Reply by tatsuz0u** [reply]
Well, it is a demo...to show the problem I mentioned can be reproduced in a normal use case.

> It should. If you use simulator, iPhone simulator use your Mac RAM. Which may not trigger the warning notification.
> 
> See source code, which explain what we did: https://github.com/SDWebImage/SDWebImage/blob/master/SDWebImage/Core/SDAnimatedImagePlayer.m#L71-L86

As you said, SD should clear memory caches if it received a memory warning.
But it didn't. I am not sure if it's a third party library issue, I am just here to prove it exists.
(If it's embedded in a LazyVStack, I guess it's more likely that the LazyVStack is to blame.)


Thanks for your tips, I'll consider using List instead or ``downsampling`` :-)

**Reply by dreampiggy** [reply]
> SD should clear memory caches if it received a memory warning.

Actually... I know this buggy behavior. Because `memory warning` (UIApplicationDidReceiveMemoryWarningNotification) is posted in the **Main Queue**

SD's animation player decode in the background queue. 😅 

There are some cases if you render too much of animation at the same time, the background queue is still submitting new request to ImageIO, and allocate frame buffer.

Because background queue does not get the main queue's notification. There need some sync solution for different queues. For example, each decoding will block to check for main queue's status whether there are already memory warning posted.

I'll find some solution later for the animation playback.

**Reply by dreampiggy** [reply]
Or some other way to detech the memory pressure, for example, [Dispatch Source's Memory Pressure](https://developer.apple.com/documentation/dispatch/dispatch_source_type_memorypressure)
